Mathlib/CategoryTheory/Yoneda.lean

```diff
@@ -504,4 +506,201 @@ lemma isIso_of_yoneda_map_bijective {X Y : C} (f : X ⟶ Y)
+  toFun φ := φ.app X.unop (ULift.up (𝟙 _))
+  invFun f :=
+    { app := fun Y x => F.map (ULift.down x) f }
+  left_inv φ := by
+    ext Y f
+    dsimp
+    rw [← FunctorToTypes.naturality]
+    dsimp
+    rw [Category.id_comp]
+    rfl
+  right_inv f := by aesop_cat
+
+/-- The Coyoneda lemma asserts that the Coyoneda pairing
+`(X : C, F : C ⥤ Type) ↦ (coyoneda.obj X ⟶ F)`
+is naturally isomorphic to the evaluation `(X, F) ↦ F.obj X`.
+
+See <https://stacks.math.columbia.edu/tag/001P>.
+-/
+def coyonedaLemma : coyonedaPairing C ≅ coyonedaEvaluation C where
+  hom :=
+    { app := fun F x => ULift.up ((x.app F.1) (𝟙 F.1))
+      naturality := by
+        intro X Y f
+        ext
+        simp [coyonedaEvaluation, ← FunctorToTypes.naturality] }
+  inv :=
+    { app := fun F x =>
+        { app := fun X a => (F.2.map a) x.down }
+      naturality := by
+        intro X Y f
+        ext
+        simp [yoneda, ← FunctorToTypes.naturality] }
+  hom_inv_id := by
+    ext
+    simp [← FunctorToTypes.naturality]
+  inv_hom_id := by
+    ext
+    simp [ULift.up_down]
+
+variable {C}
+
+/-- The isomorphism between `coyoneda.obj X ⟶ F` and `F.obj (unop X)`
+(we need to insert a `ULift` to get the universes right!)
+given by the Coyoneda lemma.
+-/
+@[simps!]
+def coyonedaSections (X : Cᵒᵖ) (F : C ⥤ Type v₁) :
+    (coyoneda.obj X ⟶ F) ≅ ULift.{u₁} (F.obj X.unop) :=
+  (coyonedaLemma C).app (X.unop, F)
+
+/-- We have a type-level equivalence between natural transformations from the coyoneda embedding
+and elements of `F.obj X.unop`, without any universe switching.
+-/
+def coyonedaEquiv {X : Cᵒᵖ} {F : C ⥤ Type v₁} : (coyoneda.obj X ⟶ F) ≃ F.obj X.unop :=
+  (coyonedaSections X F).toEquiv.trans Equiv.ulift
+
+theorem coyonedaEquiv_apply {X : Cᵒᵖ} {F : C ⥤ Type v₁} (f : coyoneda.obj X ⟶ F) :
+    coyonedaEquiv f = f.app X.unop (𝟙 X.unop) :=
+  rfl
+
+@[simp]
+theorem coyonedaEquiv_symm_app_apply {X : Cᵒᵖ} {F : C ⥤ Type v₁} (x : F.obj X.unop) (Y : C)
+    (f : X.unop ⟶ Y) : (coyonedaEquiv.symm x).app Y f = F.map f x :=
+  rfl
+
+theorem coyonedaEquiv_naturality {X Y : Cᵒᵖ} {F : C ⥤ Type v₁} (f : coyoneda.obj X ⟶ F)
+    (g : Y ⟶ X) : F.map g.unop (coyonedaEquiv f) = coyonedaEquiv (coyoneda.map g ≫ f) := by
+  change (f.app X.unop ≫ F.map g.unop) (𝟙 X.unop) = f.app Y.unop (g.unop ≫ 𝟙 Y.unop)
+  rw [← f.naturality]
+  dsimp
+  simp
+
+lemma coyonedaEquiv_naturality' {X Y : C} {F : C ⥤ Type v₁} (f : coyoneda.obj (op X) ⟶ F)
+    (g : X ⟶ Y) : F.map g (coyonedaEquiv f) = coyonedaEquiv (coyoneda.map g.op ≫ f) :=
+  coyonedaEquiv_naturality _ _
+
+lemma coyonedaEquiv_comp {X : Cᵒᵖ} {F G : C ⥤ Type v₁} (α : coyoneda.obj X ⟶ F) (β : F ⟶ G) :
+    coyonedaEquiv (α ≫ β) = β.app _ (coyonedaEquiv α) :=
+  rfl
+
+lemma coyonedaEquiv_comp' {X : C} {F G : C ⥤ Type v₁} (α : coyoneda.obj (op X) ⟶ F) (β : F ⟶ G) :
+    coyonedaEquiv (α ≫ β) = β.app X (coyonedaEquiv α) :=
+  rfl
+
+-- This lemma has always been bad, but leanprover/lean4#2644 made `simp` start noticing
```

Comment:
Do we want it as a `simp` lemma then? Removing the attribute doesn't break anything in the rest of the file.